File: src/main/java/cses/graph_algorithms/CountingRooms.java
package graph_algorithms;

import java.io.*;
import java.util.*;

public class CountingRooms {
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));

        StringTokenizer st = new StringTokenizer(reader.readLine());
        int rows = Integer.parseInt(st.nextToken());
        int columns = Integer.parseInt(st.nextToken());

        List<String> blueprint = new ArrayList<>();
        for (int i = 0; i < rows; i++) {
            blueprint.add(reader.readLine());
        }

        int answer = solution(rows, columns, blueprint);

        writer.write(answer + "\n");
        writer.flush();
    }

    static int solution(int rows, int columns, List<String> blueprint) {
        int result = 0;
        boolean[][] visited = new boolean[rows][columns];

        for (int y = 0; y < rows; y++) {
            for (int x = 0; x < columns; x++) {
                if (!visited[y][x] && blueprint.get(y).charAt(x) == '.') {
                    result++;
                    dfsIterative(rows, columns, blueprint, x, y, visited);
                }
            }
        }

        return result;
    }

    static void dfsIterative(int rows, int columns, List<String> blueprint, int startX, int startY, boolean[][] visited) {
        Deque<Point> stack = new ArrayDeque<>();
        stack.push(new Point(startX, startY));

        while (!stack.isEmpty()) {
            Point current = stack.pop();
            int x = current.x;
            int y = current.y;

            if (x < 0 || x >= columns || y < 0 || y >= rows) continue;
            if (visited[y][x]) continue;
            if (blueprint.get(y).charAt(x) != '.') continue;

            visited[y][x] = true;

            stack.push(new Point(x + 1, y));
            stack.push(new Point(x - 1, y));
            stack.push(new Point(x, y + 1));
            stack.push(new Point(x, y - 1));
        }
    }

    static class Point {
        int x, y;

        Point(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }
}

File: src/main/kotlin/graph_algorithms/CountingRooms.kt
fun main() {
//    val (n, m) = readln().split(" ").map { it.toInt() }
//    val blueprint = mutableListOf<String>()
//
//    for (i in 0..<n) {
//        blueprint.add(readln())
//    }

    val n = 5
    val m = 8
    val blueprint = listOf(
        "########",
        "#..#...#",
        "####.#.#",
        "#..#...#",
        "########"
    )

    val answer = solution(n, m, blueprint)

    println(answer)
}

fun solution(rows: Int, columns: Int, blueprint: List<String>): Int {
    var result = 0
    val visited = Array(rows) { BooleanArray(columns) { false } }
    for (y in 0 until rows) {
        for (x in 0 until columns) {
            if (!visited[y][x] && blueprint[y][x] == '.') {
                result++
                dfsIterative(rows, columns, blueprint, listOf(x, y), visited)
            }
        }
    }

    return result
}

fun dfs(rows: Int, columns: Int, blueprint: List<String>, start: List<Int>, visited: Array<BooleanArray>) {
    val (x, y) = start
    if (x < 0 || x == columns || y < 0 || y == rows) return
    if (visited[y][x]) return
    visited[y][x] = true
    if (blueprint[y][x] == '.') {
        dfs(rows, columns, blueprint, listOf(x, y + 1), visited)
        dfs(rows, columns, blueprint, listOf(x, y - 1), visited)
        dfs(rows, columns, blueprint, listOf(x + 1, y), visited)
        dfs(rows, columns, blueprint, listOf(x - 1, y), visited)
    }
}

fun dfsIterative(rows: Int, columns: Int, blueprint: List<String>, start: List<Int>, visited: Array<BooleanArray>) {
    val stack = ArrayDeque<List<Int>>()
    stack.addLast(start)

    while (stack.isNotEmpty()) {
        val (x, y) = stack.removeLast()

        if (x < 0 || x >= columns || y < 0 || y >= rows) continue
        if (visited[y][x]) continue
        if (blueprint[y][x] != '.') continue

        visited[y][x] = true

        // Add neighbors
        stack.addLast(listOf(x + 1, y))
        stack.addLast(listOf(x - 1, y))
        stack.addLast(listOf(x, y + 1))
        stack.addLast(listOf(x, y - 1))
    }
}